web-runtime/npm-package/src/FluenceAppService.ts

```diff
@@ -0,0 +1,124 @@
+import { Thread, ModuleThread, spawn, Worker } from 'threads';
+
+export const defaultNames = {
+    avm: {
+        name: 'avm.wasm',
+        package: '@fluencelabs/avm',
+    },
+    marine: {
+        name: 'marine-js.wasm',
+        package: '@fluencelabs/marine-js',
+    },
+    script: {
+        web: './marine-js.web.js',
+        node: './marine-js.node.js',
+    },
+};
+
+export const bufferToSharedArrayBuffer = (buffer: Buffer): SharedArrayBuffer => {
+    const sab = new SharedArrayBuffer(buffer.length);
+    const tmp = new Uint8Array(sab);
+    tmp.set(buffer, 0);
+    return sab;
+};
+
+export const loadWasmFromUrl = async (fileName: string): Promise<Buffer> => {
+    if (!isBrowser) {
+        throw new Error('Files can be loaded from url only in browser environment');
+    }
+    const fullUrl = window.location.origin + '/' + fileName;
+    const res = await fetch(fullUrl);
+    const ab = await res.arrayBuffer();
+    return Buffer.from(ab);
+};
+
+export const loadWasmFromNpmPackage = async (packageName: string, fileName: string): Promise<Buffer> => {
+    if (!isNode) {
+        throw new Error('Files can be loaded from npm packages only in nodejs environment');
+    }
+
+    // eval('require') is needed so that
+    // webpack will complain about missing dependencies for web target
+    const r = eval('require');
+    const path = r('path');
+    const fs = r('fs').promises;
+    const packagePath = r.resolve(packageName);
+    const filePath = path.join(path.dirname(packagePath), fileName);
+    return await fs.readFile(filePath);
+};
+
+export const loadWasm = async (args: { name: string; package: string }): Promise<SharedArrayBuffer> => {
+    let buffer: Buffer;
+    // check if we are running inside the browser and instantiate worker with the corresponding script
+    if (isBrowser) {
+        buffer = await loadWasmFromUrl(args.name);
+    }
+    // check if we are running inside nodejs and instantiate worker with the corresponding script
+    else if (isNode) {
+        buffer = await loadWasmFromNpmPackage(args.package, args.name);
+    } else {
+        throw new Error('Unknown environment');
+    }
+
+    return bufferToSharedArrayBuffer(buffer);
+};
+
+export class FluenceAppService implements IFluenceAppService {
+    private _worker?: ModuleThread<IFluenceAppService>;
+    private _workerPath: string;
+
+    constructor() {
+        // check if we are running inside the browser and instantiate worker with the corresponding script
+        if (isBrowser) {
+            this._workerPath = defaultNames.script.web;
+        }
+        // check if we are running inside nodejs and instantiate worker with the corresponding script
+        else if (isNode) {
+            this._workerPath = defaultNames.script.node;
+        } else {
+            throw new Error('Unknown environment');
+        }
+    }
+
+    async init(controlModule: SharedArrayBuffer): Promise<void> {
+        if (this._worker) {
+            return;
```

Comment:
It looks strange. Is it normal to call `init` on an already initialized object? Maybe we need to throw an exception for this case, just to catch errors as early as possible

Reply:
I wanted to have start and stop methods idempotent. The same way start and stop methods for JS Peer are implemented.

Reply:
Not sure sure if such a design can harm anything